### PR TITLE
Automatically dereference unique_ptrs in the DataBox

### PR DESCRIPTION
**Pull request by wthrowe**
The interesting stuff is in the first commit.  The fixup is changing calls to item_type and has been separated out for ease of reviewing.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: tests/Unit/DataStructures/DataBox/Test_DataBox.cpp [hunk 1]
```diff
@@ -562,6 +696,39 @@ SPECTRE_TEST_CASE("Unit.DataStructures.DataBox.apply",
...
+      original_box);
+
+  struct PointerApplyCallable {
+    using argument_tags =
+        tmpl::list<test_databox_tags::Pointer,
+                   test_databox_tags::PointerComputeItem,
+                   test_databox_tags::PointerComputeItemMutating,
+                   test_databox_tags::PointerBase,
+                   test_databox_tags::PointerComputeItemBase>;
+    static void apply(const int& s, const int& c, const int& cm, const int& sb,
```
Comment (kidder) [inline tests/Unit/DataStructures/DataBox/Test_DataBox.cpp]: please use more descriptive variable names (here and below)

**Review by kidder — CHANGES_REQUESTED** [review]
reviewed with @nilsdeppe 
You can squash everything immediately

**Comment by wthrowe** [comment]
Squashed, along with additional changes for changes on develop:
```patch
diff --git a/src/NumericalAlgorithms/Spectral/SwshDerivatives.hpp b/src/NumericalAlgorithms/Spectral/SwshDerivatives.hpp
index f60787d90..1d36725a6 100644
--- a/src/NumericalAlgorithms/Spectral/SwshDerivatives.hpp
+++ b/src/NumericalAlgorithms/Spectral/SwshDerivatives.hpp
@@ -197,7 +197,8 @@ struct AngularDerivativesImpl<tmpl::list<DerivativeTags...>,
           DerivativeTags>>*>... transform_of_derivative_scalars,
       const gsl::not_null<db::item_type<Tags::SwshTransform<
           DeDuplicatedDifferentiatedFromTags>>*>... transform_of_input_scalars,
-      const db::item_type<DeDuplicatedDifferentiatedFromTags>&... input_scalars,
+      const db::const_item_type<DeDuplicatedDifferentiatedFromTags>&...
+          input_scalars,
       const size_t l_max, const size_t number_of_radial_points) noexcept {
     apply_to_vectors(make_not_null(&get(*transform_of_derivative_scalars))...,
                      make_not_null(&get(*transform_of_input_scalars))...,
@@ -224,8 +225,8 @@ struct AngularDerivativesImpl<tmpl::list<DerivativeTags...>,
       const gsl::not_null<typename db::item_type<Tags::SwshTransform<
           DeDuplicatedDifferentiatedFromTags>>::type*>... transform_of_inputs,
       const gsl::not_null<
-          typename db::item_type<DerivativeTags>::type*>... derivatives,
-      const typename db::item_type<
+          typename db::const_item_type<DerivativeTags>::type*>... derivatives,
+      const typename db::const_item_type<
           DeDuplicatedDifferentiatedFromTags>::type&... inputs,
       const size_t l_max, const size_t number_of_radial_points) noexcept {
     // perform the forward transform on the minimal set of input nodal
```

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Comment by kidder** [comment]
ignoring Travis timeouts

**Pull request merged**